Add DFU testing workflow

szczys opened the pull request:
This PR adds a workflow to test `samples/dfu` using Twister on the nRF52840dk via a self-hosted runner.

mniestroj commented on .github/workflows/test_dfu_nrf52840dk.yml:
```diff
@@ -0,0 +1,125 @@
+        run: |
+          cd modules/lib/golioth/samples/dfu
+          west build -b nrf52840dk_nrf52840 . -p
+          west sign -t imgtool --no-hex -B build/golioth-v0.0.0.bin -- --key ../../../../../bootloader/mcuboot/root-rsa-2048.pem
+          west sign -t imgtool --no-hex -B build/golioth-v2.0.0.bin -- --key ../../../../../bootloader/mcuboot/root-rsa-2048.pem --version 2.0.0
```
```suggestion
          west sign -t imgtool --no-hex -B build/golioth-v0.0.0.bin -- --key $(west topdir)/bootloader/mcuboot/root-rsa-2048.pem
          west sign -t imgtool --no-hex -B build/golioth-v2.0.0.bin -- --key $(west topdir)/bootloader/mcuboot/root-rsa-2048.pem --version 2.0.0
```

mniestroj commented on .github/workflows/twister_dfu_nrf52840dk.yml:
```diff
@@ -0,0 +1,138 @@
+  #
+  # For item 4, the file needs to have contents like (use base64 encoding):
+  #
+  # - connected: true
+  #   id: "001050257254"
+  #   platform: nrf52840dk_nrf52840
+  #   product: J-Link
+  #   runner: nrfjprog
+  #   serial: /dev/serial/by-id/usb-SEGGER_J-Link_001050257254-if00
```
This is runner specific, so IMO it will be better to store it on the machine connected to DUT, e.g. at `/etc/hardware-map.yaml` and then mapped into the container with another volume:
```
      volumes:
        - /etc/hardware-map.yaml:/etc/hardware-map.yaml
```

szczys replied:
The OrangePis used for self-hosted runners are not easy to access. For that reason it is better to make all configuration files exist as GitHub Secrets. I have an improvement in mind that I will outline in a design document during next sprint. It will outline a process where new hardware can be added to a runner without any local configuration. I recommend going with what we have here for this PR and then refining as we review that design doc.

mniestroj replied:
Well, hardware configuration depends on what is connected to the runner. This is something that needs to be checked manually, e.g. by serial connection with Linux or with SSH. And the only way you know what is connected is by checking it physically (in order to know which DevKit/board is connected). Another thing is `/dev/serial/*` node to be used. Last thing is actually the list of connected boards (there might be many of them connected to a single runner).

I really don't see why we need to store "hardware map" (information about what is connected to a single runner) in a central database (which is GitHub Actions Secrets).

szczys replied:
It's a pain to get into the shell of these devices. They are running on their own network, so you need to connect to that network and then ssh into them. At that point the person who is running it needs to know enough about how they're set up to find the device serial endpoint and then know enough about formatting the hardware map files to create one. Then run a test using that file. It's a lot of overhead for people who don't work with these systems often. I've been working with these self-hosted runners for quite a while now and it is still a pain for me to do this.

My idea is to have a workflow that will get the output of `ls -la /dev/serial/by-id` which can then be used to create the hardware map. This way, adding new hardware means just plugging it in and running the workflow.

I think we should continue this discussion once I have submitted the design document for this.

mniestroj commented on .github/workflows/twister_dfu_nrf52840dk.yml:
```diff
@@ -0,0 +1,138 @@
+  # export GOLIOTH_SAMPLE_WIFI_SSID=golioth-runner-xxx
+  # export GOLIOTH_SAMPLE_WIFI_PSK=password_for_wifi
+  # export GOLIOTH_DEVICE_NAME=nrf52840dk
+  # export GOLIOTHCTL_CONFIG=$HOME/.goliothctl-local.yaml
```
This is not needed if the file will be located at `$HOME/.golioth/.goliothctl.yaml` (this is the default path).

mniestroj commented on .github/workflows/twister_dfu_nrf52840dk.yml:
```diff
@@ -0,0 +1,138 @@
+          sed -i 's/CONFIG_SETTINGS=y/CONFIG_SETTINGS=n/' modules/lib/golioth/samples/dfu/prj.conf
+          PATH=$PATH:/opt/nrf-command-line-tools/bin
+          nrfjprog -i
+          zephyr/scripts/twister -vvv -p nrf52840dk_nrf52840 -T modules/lib/golioth/samples/dfu --device-testing --hardware-map hardware-map.yaml --west-flash=''
```
We can provide `modules/lib/golioth/samples` to test all samples, not just `samples/dfu`. In that case it would be good to provide `-t goliothd`, as `goliothd` tag is used in all samples that require communication with Golioth backend.

szczys replied:
In recent discussion with @sam-golioth, we'd like to have cycle testing just for DFU. I'd like to merge this to focus only on the DFU sample for now and see how it performs is scheduled (an manual) runs over the next sprint, then open up a workflow that runs all samples.

sam-golioth replied:
Once we move on running all samples, we'll want each test (i.e. each sample) to show up separately in the results in Github CI, so depending on how Twister/Github spit out the results, we may want to run them all separately anyway.

mniestroj commented on .github/workflows/twister_dfu_nrf52840dk.yml:
```diff
@@ -0,0 +1,138 @@
+        run: |
+          source env_dfu_nrf52840dk.sh
+          cp .goliothctl-local.yaml ${HOME}/.
+          sed -i 's/CONFIG_SETTINGS=y/CONFIG_SETTINGS=n/' modules/lib/golioth/samples/dfu/prj.conf
```
That should not be needed anymore, at least on `main`.

mniestroj reviewed (commented):
Looks great already!

Most important thing: can we move `hardware-map.yaml` to runner?

Another (but can be part of another PR) thing is removing all the "dfu" specifics and make it more generic to run all samples, not just DFU.

szczys commented:
I added three commits to this PR that fix two issues that were causing this workflow to fail:

1. added `golioth rollback '*'` command which will rollback any existing releases before starting Twister
2. workflow adds a delay to the top of main() so logs are not missed during test.

The second point is not a permanent fix, but it will be enough for us to reliably run this test while a permanent fix is found.

mniestroj commented on .github/workflows/twister_dfu_nrf52840dk.yml:
```diff
@@ -125,6 +125,8 @@ jobs:
           PATH=$PATH:/opt/nrf-command-line-tools/bin
           nrfjprog -i
           golioth releases rollback '*'
+          # FIXME: workaround when missing the initial IDLE log
+          sed -i '/LOG_DBG("Start DFU sample");/a \\tk_sleep(K_SECONDS(10));' modules/lib/golioth/samples/dfu/src/main.c
```
How about adding `CONFIG_BOOT_DELAY=10000`? Would that help as well? We could put that into `sample.yaml` if it would work.

szczys replied:
Yes! I didn't know about this symbol, it's a much better way to handle it. Thanks.

sam-golioth commented on .github/workflows/twister_dfu_nrf52840dk.yml:
```diff
@@ -0,0 +1,139 @@
+  # Assumptions made about the self-hosted runner:
+  #
+  #  1. Has installed the GitHub Actions self-hosted runner service
+  #  2. Has nrfjprog already installed at /opt/nrf-command-line-tools
```
Just a note for posterity, we should look into options for making this managed automatically (docker, conda, etc.). As we grow the number of runners and tests, managing these dependencies and making sure they're all on the same version and such will get unwieldy.

szczys commented on .github/workflows/twister_dfu_nrf52840dk.yml:
```diff
@@ -124,6 +124,7 @@ jobs:
           source env_dfu_nrf52840dk.sh
           PATH=$PATH:/opt/nrf-command-line-tools/bin
           nrfjprog -i
+          golioth releases rollback '*'
```
I thought about that, and two things drove this decision:

* removing releases is destructive (you can't get the deleted binaries back),  but rollback can be undone.
* I think if v2.0.0 exists, we probably want the test to fail because it indicates some previous trouble that we should pay attention to

If you feel strongly about deleting instead of rolling back, let me know and I'll change it.

mniestroj replied:
There should be no other than 2.0.0 releases/artifacts, as pytest should be the only user of this project/account. Right? So even when `golioth releases rollback '*'` will disable DFU of the device, the test will still fail due to existing artifacts/releases. So the only difference it makes is that the test will fail in a different way, but it **will** fail. This is why I would either do artifacts/releases removal **or** nothing at all. Does it make sense?

mniestroj reviewed (approved):
Thanks a lot for that part! This is a big milestone, which was on my TODO list for over 2 years ;) Finally there!
